packages/preset-umi/src/features/apiRoute/apiRoute.ts

```diff
@@ -0,0 +1,144 @@
+import * as bundlerEsbuild from '@umijs/bundler-esbuild';
+import { IRoute } from '@umijs/core';
+import { logger } from '@umijs/utils';
+import fs from 'fs';
+import { join, resolve } from 'path';
+import { TEMPLATES_DIR } from '../../constants';
+import { IApi, IApiMiddleware } from '../../types';
+
+enum ServerlessPlatform {
+  Vercel = 'vercel',
+  Netlify = 'netlify',
+  Worker = 'worker',
+}
+
+function getPlatform(p: string) {
+  switch (p) {
+    case 'vercel':
+      return ServerlessPlatform.Vercel;
+    case 'netlify':
+      return ServerlessPlatform.Netlify;
+    case 'worker':
+      return ServerlessPlatform.Worker;
+    default:
+      return undefined;
+  }
+}
+
+export default (api: IApi) => {
+  let isApiRoutesGenerated = false;
+
+  // 注册 API 路由相关配置项
+  api.describe({
+    key: 'apiRoute',
+    config: {
+      schema(Joi) {
+        return Joi.object({
+          platform: Joi.string(),
+        });
+      },
+    },
+  });
+
+  // 生成中间产物时，将 API 路由与插件注册的中间件封装到临时文件目录下
+  api.onGenerateFiles(async () => {
+    const hasApiRoutes = fs.existsSync(api.paths.absSrcPath + '/api');
+    if (!hasApiRoutes) return;
+
+    const config = api.userConfig.apiRoute;
+    if (!config) {
+      logger.warn(
+        '侦测到 ./src/api 目录存在，但没有配置 config.apiRoute 选项，因此 API 路由功能将不会生效！',
+      );
+      return;
+    }
+    if (!config.platform) {
+      logger.warn(
+        '请设置 config.apiRoute.platform 配置项，否则 API 路由功能将不会生效！',
+      );
+      return;
+    }
+
+    const platform = getPlatform(config.platform);
+    if (!platform) {
+      logger.warn(
+        '配置项 config.apiRoute.platform 输入了无效的值：' +
+          config.platform +
+          '，因此 API 路由功能将不会生效！',
+      );
+      return;
+    }
+
+    if (platform !== 'vercel') {
+      logger.warn(
+        '当前版本的 Umi 仅支持将 API 路由部署到 Vercel，因此 API 路由功能将不会生效！',
+      );
+      return;
+    }
+
+    // @TODO: 根据 platform 的值执行不同 Adapter 的流程
+
+    const apiRoutes: IRoute[] = Object.keys(api.appData.apiRoutes).map(
+      (k) => api.appData.apiRoutes[k],
+    );
+
+    apiRoutes.map((apiRoute) => {
+      api.writeTmpFile({
+        noPluginDir: true,
+        path: 'api/' + apiRoute.file,
+        tplPath: join(TEMPLATES_DIR, 'apiRoute.tpl'),
+        context: {
+          adapterPath: resolve(__dirname, '../apiRoute/vercel/index.js'),
+          apiRootDirPath: api.paths.absTmpPath + '/api',
```

Comment:
```suggestion
          apiRootDirPath: join(api.paths.absTmpPath, './api'),
```

建议所有路径的地方都用 `path` 库去拼

Reply:
[8bde5c6](https://github.com/umijs/umi-next/pull/400/commits/8bde5c6c21d0a39ce911afb70f6c53902231182d) 已修复 👍